java-util/src/main/java/io/druid/java/util/common/granularity/Granularity.java

```diff
@@ -0,0 +1,538 @@
+  public static final Granularity FIFTEEN_MINUTE = fromString("FIFTEEN_MINUTE");
+  public static final Granularity THIRTY_MINUTE = fromString("THIRTY_MINUTE");
+  public static final Granularity HOUR = fromString("HOUR");
+  public static final Granularity SIX_HOUR = fromString("SIX_HOUR");
+  public static final Granularity DAY = fromString("DAY");
+  public static final Granularity WEEK = fromString("WEEK");
+  public static final Granularity MONTH = fromString("MONTH");
+  public static final Granularity QUARTER = fromString("QUARTER");
+  public static final Granularity YEAR = fromString("YEAR");
+  public static final Granularity ALL = fromString("ALL");
+  public static final Granularity NONE = fromString("NONE");
+
+  static final List<Granularity> PREDEFINED_GRANULARITIES = ImmutableList.of(
+      SECOND, MINUTE, FIVE_MINUTE, TEN_MINUTE, FIFTEEN_MINUTE, THIRTY_MINUTE,
+      HOUR, SIX_HOUR, DAY, WEEK, MONTH, QUARTER, YEAR);
+
+  // Default patterns for parsing paths.
+  final Pattern defaultPathPattern =
+      Pattern.compile(
+          "^.*[Yy]=(\\d{4})/(?:[Mm]=(\\d{2})/(?:[Dd]=(\\d{2})/(?:[Hh]=(\\d{2})/(?:[Mm]=(\\d{2})/(?:[Ss]=(\\d{2})/)?)?)?)?)?.*$"
+      );
+  final Pattern hivePathPattern =
+      Pattern.compile("^.*dt=(\\d{4})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})?)?)?)?)?)?/.*$");
+
+  @JsonCreator
+  public static Granularity fromString(String str)
+  {
+    String name = str.toUpperCase();
+    return GranularityType.createGranularity(name);
+  }
+
+  //simple merge strategy on query granularity that checks if all are equal or else
+  //returns null. this can be improved in future but is good enough for most use-cases.
+  public static Granularity mergeGranularities(List<Granularity> toMerge)
+  {
+    if (toMerge == null || toMerge.size() == 0) {
+      return null;
+    }
+
+    Granularity result = toMerge.get(0);
+    for (int i = 1; i < toMerge.size(); i++) {
+      if (!Objects.equals(result, toMerge.get(i))) {
+        return null;
+      }
+    }
+
+    return result;
+  }
+
+  public static List<Granularity> granularitiesFinerThan(final Granularity gran0)
+  {
+    final DateTime epoch = new DateTime(0);
+    final List<Granularity> retVal = Lists.newArrayList();
+    final DateTime origin = (gran0 instanceof PeriodGranularity) ? ((PeriodGranularity) gran0).getOrigin() : null;
+    final DateTimeZone tz = (gran0 instanceof PeriodGranularity) ? ((PeriodGranularity) gran0).getTimeZone() : null;
+    for (GranularityType gran : GranularityType.values()) {
+      if (gran == GranularityType.ALL || gran == GranularityType.NONE) {
+        continue;
+      }
+      final Granularity segmentGranularity = GranularityType.createGranularity(gran.name(), origin, tz);
+      if (segmentGranularity.bucket(epoch).toDurationMillis() <= gran0.bucket(epoch).toDurationMillis()) {
+        retVal.add(segmentGranularity);
+      }
+    }
+    Collections.sort(
```

Comment:
Can keep a sorted list of all granularities and make binary search + `subList()`